ICE: `Attempting to cast non-castable types FnDef(DefId(0:3 ~ mvce[19cf]::redundant), [ReErased]) and *mut u8`

<!--
[31mICE[0m: Rustc ./1C16C06CBC821A331C845B87EA1117BC52E21CEADBDEB314223270881CD88411.rs '' 'error: internal compiler error: compiler/rustc_middle/src/ty/cast.rs:99:13: Attempting to cast non-castable types FnDef(DefId(0:6 ~ 1C16C06CBC821A331C845B87EA1117BC52E21CEADBDEB314223270881CD88411[e939]::redundant), [ReErased]) and *mut u8', 'error: internal compiler error: compiler/rustc_middle/src/ty/cast.rs:99:13: Attempting to cast non-castable types FnDef(DefId(0:6 ~ 1C16C06CBC821A331C845B87EA1117BC52E21CEADBDEB314223270881CD88411[e939]::redundant), [ReErased]) and *mut u8'
File: /tmp/im/1C16C06CBC821A331C845B87EA1117BC52E21CEADBDEB314223270881CD88411.rs
-->

auto-reduced (treereduce-rust):
````rust
pub fn redundant<'a, 'a: 'a>() -> *const U {}

pub fn roundtrip() -> *const u8 {
    redundant as *mut u8 as *const u8
}


````

original:
````rust
// CHECK-NOT: as
// compile-flags: -Zinline-mir
// CHECK-LABEL: fn redundant(![crate_type = "lib"]

#[inline(always)]
fn b<'a, 'b: 'a>(x: *const T) -> *const T {
    // CHECK-LABEL: fn redundant(
    // CHECK: inlined generic_cast
    // CHECK-NOT: as
    generic_cast::<&'a u8, &'b u8>(x) as *const &'a u8
}

// EMIT_MIR casts.redundant.InstSimplify.diff
pub fn redundant<'a, 'a: 'a>(x: *mut u8) -> *const U {
    // CHECK-LABEL: fn redundant(
    // CHECK: inlined generic_cast
    // CHECK-NOT: as
    inline::<&'roundtrip u8, &'a u8>(x) as *const &'b u8
}

// EMIT_MIR casts.roundtrip.InstSimplify.diff
pub fn roundtrip(x: *const u8) -> *const u8 {
    // CHECK-LABEL: fn roundtrip(
    // CHECK: _4 = _1;
    // CHECK: _3 = move _4 as *mut u8 (PtrToPtr);
    // CHECK: _2 = move _3 as *const u8 (PointerCoercion(MutToConstPointer));
    redundant as *mut u8 as *const u8
}

````

Version information
````
rustc 1.78.0-nightly (232919c33 2024-02-10)
binary: rustc
commit-hash: 232919c33a5ba8ce8b4171b03cc898c77da136cc
commit-date: 2024-02-10
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 17.0.6

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0403]: the name `'a` is already used for a generic parameter in this item's generic parameters
 --> /tmp/icemaker_global_tempdir.OvCUtKlWIeTl/rustc_testrunner_tmpdir_reporting.FVwUTD9w2cWx/mvce.rs:1:22
  |
1 | pub fn redundant<'a, 'a: 'a>() -> *const U {}
  |                  --  ^^ already used
  |                  |
  |                  first use of `'a`

error[E0412]: cannot find type `U` in this scope
 --> /tmp/icemaker_global_tempdir.OvCUtKlWIeTl/rustc_testrunner_tmpdir_reporting.FVwUTD9w2cWx/mvce.rs:1:42
  |
1 | pub fn redundant<'a, 'a: 'a>() -> *const U {}
  |                                          ^ not found in this scope
  |
help: you might be missing a type parameter
  |
1 | pub fn redundant<'a, 'a: 'a, U>() -> *const U {}
  |                            +++

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.OvCUtKlWIeTl/rustc_testrunner_tmpdir_reporting.FVwUTD9w2cWx/mvce.rs:5:2
  |
5 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.OvCUtKlWIeTl/rustc_testrunner_tmpdir_reporting.FVwUTD9w2cWx/mvce.rs`

error: internal compiler error: compiler/rustc_middle/src/ty/cast.rs:99:13: Attempting to cast non-castable types FnDef(DefId(0:3 ~ mvce[19cf]::redundant), [ReErased]) and *mut u8

thread 'rustc' panicked at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/compiler/rustc_errors/src/lib.rs:929:30:
Box<dyn Any>
stack backtrace:
   0:     0x7f2f56d8b936 - std::backtrace_rs::backtrace::libunwind::trace::h204b59c4b7a956e8
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7f2f56d8b936 - std::backtrace_rs::backtrace::trace_unsynchronized::hedb3fd0d3233331c
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f2f56d8b936 - std::sys_common::backtrace::_print_fmt::he34d0370733de883
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f2f56d8b936 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf7ad119d240fbf03
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f2f56dde380 - core::fmt::rt::Argument::fmt::h92da3927936c4633
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/core/src/fmt/rt.rs:142:9
   5:     0x7f2f56dde380 - core::fmt::write::h0e5e470bd0ce2176
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/core/src/fmt/mod.rs:1120:17
   6:     0x7f2f56d7f17f - std::io::Write::write_fmt::heb2b1fda23254c33
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/io/mod.rs:1854:15
   7:     0x7f2f56d8b714 - std::sys_common::backtrace::_print::h263853e0ea367314
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f2f56d8b714 - std::sys_common::backtrace::print::h00f9d494ba8e0e79
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f2f56d8e4d7 - std::panicking::default_hook::{{closure}}::hacb2803bd6938000
  10:     0x7f2f56d8e239 - std::panicking::default_hook::hb716af7cb4653a24
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/panicking.rs:292:9
  11:     0x7f2f59b29d8c - std[7f0fb7d071667f83]::panicking::update_hook::<alloc[ec204197b905d5dc]::boxed::Box<rustc_driver_impl[4971e3aefab291b7]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f2f56d8ec26 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h1f9f88256f710aa3
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/alloc/src/boxed.rs:2030:9
  13:     0x7f2f56d8ec26 - std::panicking::rust_panic_with_hook::h28bab546a3573577
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/panicking.rs:785:13
  14:     0x7f2f59b5b294 - std[7f0fb7d071667f83]::panicking::begin_panic::<rustc_errors[4ebaeed1ac72940a]::ExplicitBug>::{closure#0}
  15:     0x7f2f59b57986 - std[7f0fb7d071667f83]::sys_common::backtrace::__rust_end_short_backtrace::<std[7f0fb7d071667f83]::panicking::begin_panic<rustc_errors[4ebaeed1ac72940a]::ExplicitBug>::{closure#0}, !>
  16:     0x7f2f59b530b6 - std[7f0fb7d071667f83]::panicking::begin_panic::<rustc_errors[4ebaeed1ac72940a]::ExplicitBug>
  17:     0x7f2f59b664d1 - <rustc_errors[4ebaeed1ac72940a]::diagnostic_builder::BugAbort as rustc_errors[4ebaeed1ac72940a]::diagnostic_builder::EmissionGuarantee>::emit_producing_guarantee
  18:     0x7f2f59f1c61e - <rustc_errors[4ebaeed1ac72940a]::DiagCtxt>::bug::<alloc[ec204197b905d5dc]::string::String>
  19:     0x7f2f59fb741b - rustc_middle[633bf8955aaf4eba]::util::bug::opt_span_bug_fmt::<rustc_span[6c817caf5980a8f7]::span_encoding::Span>::{closure#0}
  20:     0x7f2f59f9e85a - rustc_middle[633bf8955aaf4eba]::ty::context::tls::with_opt::<rustc_middle[633bf8955aaf4eba]::util::bug::opt_span_bug_fmt<rustc_span[6c817caf5980a8f7]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f2f59f9e6d8 - rustc_middle[633bf8955aaf4eba]::ty::context::tls::with_context_opt::<rustc_middle[633bf8955aaf4eba]::ty::context::tls::with_opt<rustc_middle[633bf8955aaf4eba]::util::bug::opt_span_bug_fmt<rustc_span[6c817caf5980a8f7]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f2f57f80b30 - rustc_middle[633bf8955aaf4eba]::util::bug::bug_fmt
  23:     0x7f2f5c1055ce - rustc_middle[633bf8955aaf4eba]::ty::cast::mir_cast_kind.cold.0
  24:     0x7f2f5b25d9e1 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_rvalue
  25:     0x7f2f5b57b335 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::expr_into_dest
  26:     0x7f2f5aeb1296 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_temp::{closure#0}
  27:     0x7f2f5ae598c4 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_operand
  28:     0x7f2f5b25c0c7 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_rvalue
  29:     0x7f2f5b57b335 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::expr_into_dest
  30:     0x7f2f5aeb1296 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_temp::{closure#0}
  31:     0x7f2f5ae598c4 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_operand
  32:     0x7f2f5ae5958f - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_operand
  33:     0x7f2f5b25d98b - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::as_rvalue
  34:     0x7f2f5b57b335 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::expr_into_dest
  35:     0x7f2f5b57d4c5 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::expr_into_dest
  36:     0x7f2f5813e75d - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::ast_block_stmts
  37:     0x7f2f5b57c198 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::expr_into_dest
  38:     0x7f2f5b57d4c5 - <rustc_mir_build[9635c1b0f9da976d]::build::Builder>::expr_into_dest
  39:     0x7f2f5b265ae9 - rustc_mir_build[9635c1b0f9da976d]::build::mir_build::{closure#0}
  40:     0x7f2f5b261e6b - rustc_mir_build[9635c1b0f9da976d]::build::mir_built
  41:     0x7f2f5b261b55 - rustc_query_impl[d07a5c1abde99c62]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d07a5c1abde99c62]::query_impl::mir_built::dynamic_query::{closure#2}::{closure#0}, rustc_middle[633bf8955aaf4eba]::query::erase::Erased<[u8; 8usize]>>
  42:     0x7f2f5af7b33a - rustc_query_system[41c20884e3c50d50]::query::plumbing::try_execute_query::<rustc_query_impl[d07a5c1abde99c62]::DynamicConfig<rustc_query_system[41c20884e3c50d50]::query::caches::VecCache<rustc_span[6c817caf5980a8f7]::def_id::LocalDefId, rustc_middle[633bf8955aaf4eba]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[d07a5c1abde99c62]::plumbing::QueryCtxt, false>
  43:     0x7f2f5af7af10 - rustc_query_impl[d07a5c1abde99c62]::query_impl::mir_built::get_query_non_incr::__rust_end_short_backtrace
  44:     0x7f2f5b1c0b21 - rustc_mir_build[9635c1b0f9da976d]::check_unsafety::check_unsafety
  45:     0x7f2f5b1c0949 - rustc_query_impl[d07a5c1abde99c62]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d07a5c1abde99c62]::query_impl::check_unsafety::dynamic_query::{closure#2}::{closure#0}, rustc_middle[633bf8955aaf4eba]::query::erase::Erased<[u8; 0usize]>>
  46:     0x7f2f5b6812cd - rustc_query_system[41c20884e3c50d50]::query::plumbing::try_execute_query::<rustc_query_impl[d07a5c1abde99c62]::DynamicConfig<rustc_query_system[41c20884e3c50d50]::query::caches::VecCache<rustc_span[6c817caf5980a8f7]::def_id::LocalDefId, rustc_middle[633bf8955aaf4eba]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[d07a5c1abde99c62]::plumbing::QueryCtxt, false>
  47:     0x7f2f5b681017 - rustc_query_impl[d07a5c1abde99c62]::query_impl::check_unsafety::get_query_non_incr::__rust_end_short_backtrace
  48:     0x7f2f5b895086 - rustc_interface[2314d7060e0f18dd]::passes::analysis
  49:     0x7f2f5b894adf - rustc_query_impl[d07a5c1abde99c62]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[d07a5c1abde99c62]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[633bf8955aaf4eba]::query::erase::Erased<[u8; 1usize]>>
  50:     0x7f2f5bb28572 - rustc_query_system[41c20884e3c50d50]::query::plumbing::try_execute_query::<rustc_query_impl[d07a5c1abde99c62]::DynamicConfig<rustc_query_system[41c20884e3c50d50]::query::caches::SingleCache<rustc_middle[633bf8955aaf4eba]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[d07a5c1abde99c62]::plumbing::QueryCtxt, false>
  51:     0x7f2f5bb282d5 - rustc_query_impl[d07a5c1abde99c62]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  52:     0x7f2f5bb3e7c2 - rustc_interface[2314d7060e0f18dd]::interface::run_compiler::<core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>, rustc_driver_impl[4971e3aefab291b7]::run_compiler::{closure#0}>::{closure#0}
  53:     0x7f2f5bbdc806 - std[7f0fb7d071667f83]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[2314d7060e0f18dd]::util::run_in_thread_with_globals<rustc_interface[2314d7060e0f18dd]::util::run_in_thread_pool_with_globals<rustc_interface[2314d7060e0f18dd]::interface::run_compiler<core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>, rustc_driver_impl[4971e3aefab291b7]::run_compiler::{closure#0}>::{closure#0}, core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>>::{closure#0}, core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>>
  54:     0x7f2f5bbdc633 - <<std[7f0fb7d071667f83]::thread::Builder>::spawn_unchecked_<rustc_interface[2314d7060e0f18dd]::util::run_in_thread_with_globals<rustc_interface[2314d7060e0f18dd]::util::run_in_thread_pool_with_globals<rustc_interface[2314d7060e0f18dd]::interface::run_compiler<core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>, rustc_driver_impl[4971e3aefab291b7]::run_compiler::{closure#0}>::{closure#0}, core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>>::{closure#0}, core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[b966835390a286e1]::result::Result<(), rustc_span[6c817caf5980a8f7]::ErrorGuaranteed>>::{closure#1} as core[b966835390a286e1]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  55:     0x7f2f56d97eb5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha3c43b6c892c82cf
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/alloc/src/boxed.rs:2016:9
  56:     0x7f2f56d97eb5 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h39babcbf5760ace1
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/alloc/src/boxed.rs:2016:9
  57:     0x7f2f56d97eb5 - std::sys::pal::unix::thread::Thread::new::thread_start::h35dde8f54f904bd0
                               at /rustc/232919c33a5ba8ce8b4171b03cc898c77da136cc/library/std/src/sys/pal/unix/thread.rs:108:17
  58:     0x7f2f56b4b9eb - <unknown>
  59:     0x7f2f56bcf7cc - <unknown>
  60:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.78.0-nightly (232919c33 2024-02-10) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [mir_built] building MIR for `roundtrip`
#1 [check_unsafety] unsafety-checking `roundtrip`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 4 previous errors

Some errors have detailed explanations: E0403, E0412, E0601.
For more information about an error, try `rustc --explain E0403`.

```

</p>
</details>

<!--
query stack:
#0 [mir_built] building MIR for `roundtrip`
#1 [check_unsafety] unsafety-checking `roundtrip`
#2 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -186,17 +186,6 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
         let b = self.shallow_resolve(b);
         debug!("Coerce.tys({:?} => {:?})", a, b);
 
-        // Just ignore error types.
-        if let Err(guar) = (a, b).error_reported() {
-            // Best-effort try to unify these types -- we're already on the error path,
-            // so this will have the side-effect of making sure we have no ambiguities
-            // due to `[type error]` and `_` not coercing together.
-            let _ = self.commit_if_ok(|_| {
-                self.at(&self.cause, self.param_env).eq(DefineOpaqueTypes::Yes, a, b)
-            });
-            return success(vec![], Ty::new_error(self.fcx.tcx, guar), vec![]);
-        }
-
         // Coercing from `!` to any type is allowed:
         if a.is_never() {
             return success(simple(Adjust::NeverToAny)(b), b, vec![]);

--- a/tests/ui/coercion/type-errors.rs
+++ b/tests/ui/coercion/type-errors.rs
@@ -1,0 +1,15 @@
+// Regression test for an ICE: https://github.com/rust-lang/rust/issues/120884
+// We still need to properly go through coercions between types with errors instead of
+// shortcutting and returning success, because we need the adjustments for building the MIR.
+
+pub fn has_error() -> TypeError {}
+//~^ ERROR cannot find type `TypeError` in this scope
+
+pub fn cast() -> *const u8 {
+    // Casting a function item to a data pointer in valid in HIR, but invalid in MIR.
+    // We need an adjustment (ReifyFnPointer) to insert a cast from the function item
+    // to a function pointer as a separate MIR statement.
+    has_error as *const u8
+}
+
+fn main() {}

--- a/tests/ui/coercion/type-errors.stderr
+++ b/tests/ui/coercion/type-errors.stderr
@@ -1,0 +1,9 @@
+error[E0412]: cannot find type `TypeError` in this scope
+  --> $DIR/type-errors.rs:5:23
+   |
+LL | pub fn has_error() -> TypeError {}
+   |                       ^^^^^^^^^ not found in this scope
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0412`.

--- a/tests/ui/error-codes/E0401.rs
+++ b/tests/ui/error-codes/E0401.rs
@@ -9,8 +9,6 @@ fn foo<T>(x: T) {
            (y: T) { //~ ERROR E0401
     }
     bfnr(x);
-    //~^ ERROR type annotations needed
-    //~| ERROR type annotations needed
 }
 
 

--- a/tests/ui/error-codes/E0401.stderr
+++ b/tests/ui/error-codes/E0401.stderr
@@ -21,7 +21,7 @@ LL |            (y: T) {
    |                ^ use of generic parameter from outer item
 
 error[E0401]: can't use `Self` from outer item
-  --> $DIR/E0401.rs:24:25
+  --> $DIR/E0401.rs:22:25
    |
 LL | impl<T> Iterator for A<T> {
    | ---- `Self` type implicitly declared here, by this `impl`
@@ -32,45 +32,6 @@ LL |         fn helper(sel: &Self) -> u8 {
    |                         use of `Self` from outer item
    |                         refer to the type directly here instead
 
-error[E0283]: type annotations needed
-  --> $DIR/E0401.rs:11:5
-   |
-LL |     bfnr(x);
-   |     ^^^^ cannot infer type of the type parameter `V` declared on the function `bfnr`
-   |
-   = note: cannot satisfy `_: Baz<_>`
-note: required by a bound in `bfnr`
-  --> $DIR/E0401.rs:4:19
-   |
-LL |     fn bfnr<U, V: Baz<U>, W: Fn()>(y: T) {
-   |                   ^^^^^^ required by this bound in `bfnr`
-help: consider specifying the generic arguments
-   |
-LL |     bfnr::<U, V, W>(x);
-   |         +++++++++++
+error: aborting due to 3 previous errors
 
-error[E0283]: type annotations needed
-  --> $DIR/E0401.rs:11:5
-   |
-LL |     bfnr(x);
-   |     ^^^^ cannot infer type of the type parameter `W` declared on the function `bfnr`
-   |
-   = note: multiple `impl`s satisfying `_: Fn()` found in the following crates: `alloc`, `core`:
-           - impl<A, F> Fn<A> for &F
-             where A: Tuple, F: Fn<A>, F: ?Sized;
-           - impl<Args, F, A> Fn<Args> for Box<F, A>
-             where Args: Tuple, F: Fn<Args>, A: Allocator, F: ?Sized;
-note: required by a bound in `bfnr`
-  --> $DIR/E0401.rs:4:30
-   |
-LL |     fn bfnr<U, V: Baz<U>, W: Fn()>(y: T) {
-   |                              ^^^^ required by this bound in `bfnr`
-help: consider specifying the generic arguments
-   |
-LL |     bfnr::<U, V, W>(x);
-   |         +++++++++++
-
-error: aborting due to 5 previous errors
-
-Some errors have detailed explanations: E0283, E0401.
-For more information about an error, try `rustc --explain E0283`.
+For more information about this error, try `rustc --explain E0401`.

--- a/tests/ui/typeck/issue-104510-ice.rs
+++ b/tests/ui/typeck/issue-104510-ice.rs
@@ -6,7 +6,7 @@ struct W<T: ?Sized>(Oops);
 
 unsafe fn test() {
     let j = W(());
-    let pointer = &j as *const _; //~ ERROR type annotations needed
+    let pointer = &j as *const _;
     core::arch::asm!(
         "nop",
         in("eax") pointer,

--- a/tests/ui/typeck/issue-104510-ice.stderr
+++ b/tests/ui/typeck/issue-104510-ice.stderr
@@ -4,18 +4,6 @@ error[E0412]: cannot find type `Oops` in this scope
 LL | struct W<T: ?Sized>(Oops);
    |                     ^^^^ not found in this scope
 
-error[E0282]: type annotations needed for `*const W<T>`
-  --> $DIR/issue-104510-ice.rs:9:9
-   |
-LL |     let pointer = &j as *const _;
-   |         ^^^^^^^
-   |
-help: consider giving `pointer` an explicit type, where the type for type parameter `T` is specified
-   |
-LL |     let pointer: *const W<T> = &j as *const _;
-   |                +++++++++++++
+error: aborting due to 1 previous error
 
-error: aborting due to 2 previous errors
-
-Some errors have detailed explanations: E0282, E0412.
-For more information about an error, try `rustc --explain E0282`.
+For more information about this error, try `rustc --explain E0412`.

